src/device_trezor/trezor/protocol.cpp

```diff
@@ -0,0 +1,1074 @@
+    return a->spend_public_key() == b->spend_public_key() && a->view_public_key() == b->view_public_key();
+  }
+
+  std::string hash_addr(const MoneroAccountPublicAddress * addr, boost::optional<uint64_t> amount, boost::optional<bool> is_subaddr){
+    return hash_addr(addr->spend_public_key(), addr->view_public_key(), amount, is_subaddr);
+  }
+
+  std::string hash_addr(const std::string & spend_key, const std::string & view_key, boost::optional<uint64_t> amount, boost::optional<bool> is_subaddr){
+    ::crypto::public_key spend{}, view{};
+    if (spend_key.size() != 32 || view_key.size() != 32){
+      throw std::invalid_argument("Public keys have invalid sizes");
+    }
+
+    memcpy(spend.data, spend_key.data(), 32);
+    memcpy(view.data, view_key.data(), 32);
+    return hash_addr(&spend, &view, amount, is_subaddr);
+  }
+
+  std::string hash_addr(const ::crypto::public_key * spend_key, const ::crypto::public_key * view_key, boost::optional<uint64_t> amount, boost::optional<bool> is_subaddr){
+    char buff[64+8+1];
+    size_t offset = 0;
+
+    memcpy(buff + offset, spend_key->data, 32); offset += 32;
+    memcpy(buff + offset, view_key->data, 32); offset += 32;
+
+    if (amount){
+      memcpy(buff + offset, (uint8_t*) &(amount.get()), sizeof(amount.get())); offset += sizeof(amount.get());
+    }
+
+    if (is_subaddr){
+      buff[offset] = is_subaddr.get();
+      offset += 1;
+    }
+
+    return std::string(buff, offset);
+  }
+
+  TData::TData() {
+    in_memory = false;
+    rsig_type = 0;
+    cur_input_idx = 0;
+    cur_output_idx = 0;
+    cur_batch_idx = 0;
+    cur_output_in_batch_idx = 0;
+  }
+
+  Signer::Signer(wallet_shim *wallet2, const unsigned_tx_set * unsigned_tx, size_t tx_idx, hw::tx_aux_data * aux_data) {
+    m_wallet2 = wallet2;
+    m_unsigned_tx = unsigned_tx;
+    m_aux_data = aux_data;
+    m_tx_idx = tx_idx;
+    m_ct.tx_data = cur_tx();
+    m_multisig = false;
+  }
+
+  void Signer::extract_payment_id(){
+    const std::vector<uint8_t>& tx_extra = cur_tx().extra;
+    m_ct.tsx_data.set_payment_id("");
+
+    std::vector<cryptonote::tx_extra_field> tx_extra_fields;
+    cryptonote::parse_tx_extra(tx_extra, tx_extra_fields); // ok if partially parsed
+    cryptonote::tx_extra_nonce extra_nonce;
+
+    ::crypto::hash payment_id{};
+    if (find_tx_extra_field_by_type(tx_extra_fields, extra_nonce))
+    {
+      ::crypto::hash8 payment_id8{};
+      if(cryptonote::get_encrypted_payment_id_from_tx_extra_nonce(extra_nonce.nonce, payment_id8))
+      {
+        m_ct.tsx_data.set_payment_id(std::string(payment_id8.data, 8));
+      }
+      else if (cryptonote::get_payment_id_from_tx_extra_nonce(extra_nonce.nonce, payment_id))
+      {
+        m_ct.tsx_data.set_payment_id(std::string(payment_id.data, 32));
+      }
+    }
+  }
+
+  static unsigned get_rsig_type(bool use_bulletproof, size_t num_outputs){
+    if (!use_bulletproof){
+      return 0;  // Borromean
+    } else if (num_outputs > 16){
```

Comment:
You defined a macro with this earlier, and don't even use it :P 
It's in cryptonote_config.h though.